File: src/main/java/com/fish/list/Q25.java
package com.fish.list;

import org.junit.jupiter.api.Test;

/**
 * @author liuqi
 * @date 2023/11/20
 */
public class Q25 {
    class Solution {
        public ListNode reverseKGroup(ListNode head, int k) {
            ListNode first = head;
            ListNode last = head;
            ListNode pre = new ListNode();
            ListNode newHead = null;
            pre.next = head;
            while (true) {
                // 往后找k个
                for (int i = 1; i < k; i++) {
                    if (last == null) {
                        break;
                    }
                    last = last.next;
                }
                if (last != null && newHead == null) {
                    newHead = last;
                }
                if (last == null) {
                    return newHead == null ? head : newHead;
                }
                // 循环k-1次
                ListNode oldF = first;
                for (int i = 1; i < k; i++) {
                    ListNode temp = first.next;
                    first.next = last.next;
                    last.next = first;
                    first = temp;
                }
                pre.next = last;
                pre = oldF;
                first = last = oldF.next;
            }
        }
    }

    @Test
    public void test() {
        ListNode head = new ListNode(1);
        ListNode t = head;
        for (int i = 0; i < 4; i++) {
            t.next = new ListNode(i + 2);
            t = t.next;
        }
        new Solution().reverseKGroup(head, 2);
    }
}

File: python/Q25.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    def reverseList(self, head, k):
        if not head:
            return head, None
        # 往后遍历，判断是否足够len个节点
        cnt = 0
        temp = head
        while cnt < k:
            cnt += 1
            if not temp:
                return head, None
            temp = temp.next
        # 开始进行翻转
        pre = head
        curr = head.next
        pre.next = None
        while curr != temp:
            tt = curr.next
            curr.next = pre
            pre = curr
            curr = tt
        return pre, temp

    def reverseKGroup(self, head, k):
        now = head
        newHead = None
        preLast = ListNode(0, None)
        while now:
            # 尝试翻转后面的连续k个节点，返回翻转后的头节点
            preLast.next, nextNow = self.reverseList(now, k)
            if not newHead:
                newHead = preLast.next
            # 如果翻转完毕，则直接返回结果
            if not nextNow:
                return newHead
            # 更新指针
            preLast = now
            now = nextNow
        return newHead
